examples/analytics-3p-remote-frame-helper.js

```diff
@@ -0,0 +1,188 @@
+  constructor() {
+    /**
+     * @type {Function<!String>}
+     * @private
+     */
+    this.listener_ = () => {};
+
+    /**
+     * @type {Function<!String>}
+     * @private
+     */
+    this.extraDataListener_ = () => {};
+
+    /**
+     * Holds a mapping between sender ID and extra data
+     * @type {Map}
+     * @private
+     */
+    this.senderIdToExtraData_ = {}; // TODO: Replace with map from AMP
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!Function} listener A function that takes an array of event
+   * strings, and does something with them.
+   * @param {?Function} extraDataListener A function to receive any extra
+   * setup data sent by the creative
+   */
+  registerAmpAnalyticsEventListener(listener, extraDataListener) {
+    this.listener_ = listener;
+    this.extraDataListener_ = extraDataListener;
+  }
+
+  /**
+   * Receives a message from the creative. Handles deserialization (which
+   * filters out messages not actually meant for us) and passes it to the
+   * listener that was supplied to registerAmpAnalyticsEventListener above.
+   * @param {!String} message The message that was received.
+   */
+  receiveMessageFromCreative(message) {
+    if (message && message.data) {
+      const deserialized = deserializeMessage(message.data);
+      if (deserialized) {
+        this.setSentinel_(deserialized.sentinel);
+        if (deserialized.ampAnalyticsExtraData) {
+          this.senderIdToExtraData_[deserialized.senderId] =
+            deserialized.ampAnalyticsExtraData;
+          window.requestIdleCallback(() => {
+            this.extraDataListener_(deserialized.senderId,
+              deserialized.ampAnalyticsExtraData);
+          });
+        } else if (deserialized.ampAnalyticsEvents) {
+          window.requestIdleCallback(() => {
+            this.listener_(deserialized.ampAnalyticsEvents);
+          });
+        }
+      }
+    }
+  }
+
+  /**
+   * Sets the sentinel value, which is used to identify which creative the
+   * event messages come from
+   * @param {!String} sentinel The sentinel value
+   */
+  setSentinel_(sentinel) {
+    if (this.sentinel_ && sentinel != this.sentinel_) {
+      console.warn("Attempting to set sentinel to " + sentinel + " when it" +
+          " was already set to " + this.sentinel_);
+    }
+    this.sentinel_ = sentinel;
+  }
+
+  /**
+   * Gets any optional extra data that should be made available to the
+   * cross-domain frame, in the context of a particular creative.
+   * @param {!number} senderId The ID of the creative that sent the extra data
+   * @returns {String=}
+   */
+  getExtraData(senderId) {
+    return this.senderIdToExtraData_[senderId];
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @param {!String} message The message to send.
+   */
+  sendMessageToCreative(message) {
+    /*REVIEW*/window.parent.postMessage(
```

Comment:
* For messages from amp-analytics to xframe: yes, using the iframeMessagingClient abstraction
* For responses from xframe to amp-analytics: didn't, because haven't done AMP imports yet, but can do.